js/services.ts

```diff
@@ -139,12 +143,127 @@ module jsonforms.services {
+                "required": this.keys(this.requiredProperties(properties))
+            };
+        };
+
+        private properties = (instance: Object) : Object => {
+            var properties = {};
+            var generator = this;
+            this.keys(instance).forEach(function(property) {
+                properties[property] = generator.property(instance[property])
+            });
+            return properties;
+        };
+
+        private keys = (properties: Object) : string[] => {
+            return Object.keys(properties);
+        };
+
+        private property = (instance: any) : Object => {
+            switch (typeof instance) {
+                case "string":
+                case "boolean":
+                    return { "type": typeof instance };
+                case "number":
+                    if (Number(instance) % 1 === 0) {
+                        return { "type": "integer" };
+                    } else {
+                        return { "type": "number" };
+                    }
+                case "object":
+                    return this.schemaObjectOrNullOrArray(instance);
+                default:
+                    return {};
+            }
+        };
+
+        private schemaObjectOrNullOrArray= (instance: Object): Object => {
+            if (this.isNotNull(instance)) {
+                if (this.isArray(instance)) {
+                    return this.schemaArray(<Array<Object>>instance);
+                } else {
+                    return this.schemaObject(instance);
+                }
+            } else {
+                return { "type": "null" };
+            }
+        };
+
+        private schemaArray= (instance: Array<Object>): Object => {
+            if (instance.length) {
+                var generator = this;
+                var allProperties = instance.map(function(object) {
+                    return generator.property(object);
+                });
+                var uniqueProperties = this.distinct(allProperties,
+                    function(object) { return JSON.stringify(object) });
+                if (uniqueProperties.length == 1) {
+                    return {
+                        "type": "array",
+                        "items": uniqueProperties[0]
+                    };
+                } else {
+                    return {
+                        "type": "array",
+                        "items": {
+                            "oneOf": uniqueProperties
+                        }
+                    };
+                }
+            }
+        };
+
+        private isArray = (instance: any): boolean => {
+            return Object.prototype.toString.call(instance) === '[object Array]';
+        };
+
+        private isNotNull = (instance: any): boolean => {
+            return (typeof(instance) !== 'undefined') && (instance !== null);
+        };
+
+        private distinct = (array: Array<Object>, key: (item: Object) => any): Array<Object> => {
```

Comment:
I'd rename key to something else. It looks more like a transformation function that determines uniqueness. I makes sense when looking at the implementation of distinct, but I'd guess it doesn't matter for any callers.